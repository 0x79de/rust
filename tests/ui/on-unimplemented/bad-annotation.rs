#![crate_type = "lib"]
#![feature(rustc_attrs)]
#![allow(unused)]

#[rustc_on_unimplemented = "test error `{Self}` with `{Bar}` `{Baz}` `{Quux}`"]
trait Foo<Bar, Baz, Quux> {}

#[rustc_on_unimplemented = "a collection of type `{Self}` cannot \
 be built from an iterator over elements of type `{A}`"]
trait MyFromIterator<A> {
    /// Builds a container with elements from an external iterator.
    fn my_from_iter<T: Iterator<Item = A>>(iterator: T) -> Self;
}

#[rustc_on_unimplemented]
//~^ ERROR malformed `rustc_on_unimplemented` attribute
trait NoContent {}

#[rustc_on_unimplemented = "Unimplemented trait error on `{Self}` with params `<{A},{B},{C}>`"]
//~^ ERROR cannot find parameter C on this trait
trait ParameterNotPresent<A, B> {}

#[rustc_on_unimplemented = "Unimplemented trait error on `{Self}` with params `<{A},{B},{}>`"]
//~^ ERROR positional format arguments are not allowed here
trait NoPositionalArgs<A, B> {}

#[rustc_on_unimplemented(lorem = "")]
//~^ ERROR this attribute must have a valid
trait EmptyMessage {}

#[rustc_on_unimplemented(lorem(ipsum(dolor)))]
//~^ ERROR this attribute must have a valid
trait Invalid {}

#[rustc_on_unimplemented(message = "x", message = "y")]
//~^ ERROR this attribute must have a valid
trait DuplicateMessage {}

#[rustc_on_unimplemented(message = "x", on(desugared, message = "y"))]
//~^ ERROR this attribute must have a valid
trait OnInWrongPosition {}

#[rustc_on_unimplemented(on(), message = "y")]
//~^ ERROR empty `on`-clause
trait NoEmptyOn {}

#[rustc_on_unimplemented(on = "x", message = "y")]
//~^ ERROR this attribute must have a valid
trait ExpectedPredicateInOn {}

#[rustc_on_unimplemented(on(x = "y"), message = "y")]
trait OnWithoutDirectives {}

#[rustc_on_unimplemented(on(desugared, on(desugared, message = "x")), message = "y")]
//~^ ERROR this attribute must have a valid
trait NoNestedOn {}

// caught by `OnUnimplementedDirective::parse`, *not* `eval_condition`
#[rustc_on_unimplemented(on("y", message = "y"))]
//~^ ERROR invalid `on`-clause
trait UnsupportedLiteral {}

#[rustc_on_unimplemented(on(not(a, b), message = "y"))]
//~^ ERROR expected 1 cfg-pattern
trait ExpectedOnePattern {}

#[rustc_on_unimplemented(on(thing::What, message = "y"))]
//~^ ERROR `cfg` predicate key must be an identifier
trait KeyMustBeIdentifier {}

#[rustc_on_unimplemented(on(thing::What = "value", message = "y"))]
//~^ ERROR `cfg` predicate key must be an identifier
trait KeyMustBeIdentifier2 {}
