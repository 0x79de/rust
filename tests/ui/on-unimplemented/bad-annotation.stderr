error: malformed `rustc_on_unimplemented` attribute input
  --> $DIR/bad-annotation.rs:17:1
   |
LL | #[rustc_on_unimplemented]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: the following are the possible correct uses
   |
LL | #[rustc_on_unimplemented = "message"]
   |                          +++++++++++
LL | #[rustc_on_unimplemented(/*opt*/ message = "...", /*opt*/ label = "...", /*opt*/ note = "...")]
   |                         ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

warning: there is no parameter `C` on trait `BadAnnotation2`
  --> $DIR/bad-annotation.rs:22:90
   |
LL | #[rustc_on_unimplemented = "Unimplemented trait error on `{Self}` with params `<{A},{B},{C}>`"]
   |                                                                                          ^
   |
   = help: expect either a generic argument name or `{Self}` as format argument
   = note: `#[warn(unknown_or_malformed_diagnostic_attributes)]` on by default

warning: positional format arguments are not allowed here
  --> $DIR/bad-annotation.rs:27:90
   |
LL | #[rustc_on_unimplemented = "Unimplemented trait error on `{Self}` with params `<{A},{B},{}>`"]
   |                                                                                          ^
   |
   = help: only named format arguments with the name of one of the generic types are allowed in this context

error[E0232]: this attribute must have a valid value
  --> $DIR/bad-annotation.rs:32:26
   |
LL | #[rustc_on_unimplemented(lorem="")]
   |                          ^^^^^^^^ expected value here
   |
   = note: eg `#[rustc_on_unimplemented(message="foo")]`

error[E0232]: this attribute must have a valid value
  --> $DIR/bad-annotation.rs:36:26
   |
LL | #[rustc_on_unimplemented(lorem(ipsum(dolor)))]
   |                          ^^^^^^^^^^^^^^^^^^^ expected value here
   |
   = note: eg `#[rustc_on_unimplemented(message="foo")]`

error[E0232]: this attribute must have a valid value
  --> $DIR/bad-annotation.rs:40:39
   |
LL | #[rustc_on_unimplemented(message="x", message="y")]
   |                                       ^^^^^^^^^^^ expected value here
   |
   = note: eg `#[rustc_on_unimplemented(message="foo")]`

error[E0232]: this attribute must have a valid value
  --> $DIR/bad-annotation.rs:44:39
   |
LL | #[rustc_on_unimplemented(message="x", on(desugared, message="y"))]
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected value here
   |
   = note: eg `#[rustc_on_unimplemented(message="foo")]`

error[E0232]: empty `on`-clause in `#[rustc_on_unimplemented]`
  --> $DIR/bad-annotation.rs:48:26
   |
LL | #[rustc_on_unimplemented(on(), message="y")]
   |                          ^^^^ empty on-clause here

error[E0232]: this attribute must have a valid value
  --> $DIR/bad-annotation.rs:52:26
   |
LL | #[rustc_on_unimplemented(on="x", message="y")]
   |                          ^^^^^^ expected value here
   |
   = note: eg `#[rustc_on_unimplemented(message="foo")]`

error[E0232]: this attribute must have a valid value
  --> $DIR/bad-annotation.rs:59:40
   |
LL | #[rustc_on_unimplemented(on(desugared, on(desugared, message="x")), message="y")]
   |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected value here
   |
   = note: eg `#[rustc_on_unimplemented(message="foo")]`

error: aborting due to 8 previous errors; 2 warnings emitted

For more information about this error, try `rustc --explain E0232`.
